### PR TITLE
Bug 1164341 - [Stingray][Dashboard] Clock widget skeleton

### PR DESCRIPTION
**Pull request by rexboy7**

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fa1c6f6424e260eeccc9a487e9dbdc588d8a65fa

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=17c504b462fd0321bebeffebbd496d20eb2977dd

File: build/config/tv/settings.json [hunk 1]
```diff
@@ -1,4 +1,4 @@
 {
  "homescreen.appName": "smart-home",
- "landing_app.manifestURL": "app://clock.gaiamobile.org/manifest.webapp"
+ "landing_app.manifestURL": "app://dashboard.gaiamobile.org/manifest.webapp"
```
Comment (huchengtw-moz) [inline build/config/tv/settings.json]: Nice to see this. Thanks.

File: tv_apps/smart-clock/manifest.webapp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "Smart Clock",
+  "type": "certified",
+  "role": "widget",
```
Comment (huchengtw-moz) [inline tv_apps/smart-clock/manifest.webapp]: There is no role called `widget`. We should change it to `system` or modify app-deck to skip this role.

File: tv_apps/smart-clock/js/clock_component.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    return component;
+  },
+
+  updateAnalogClock: function cc_updateAnalogClock() {
+    var now = new Date();
+    var min, hour;
+    min = now.getMinutes();
+    // hours progress gradually
+    hour = (now.getHours() % 12) + min / 60;
+    this.setTransform('minute', 34);
```
Comment (huchengtw-moz) [inline tv_apps/smart-clock/js/clock_component.js]: use `min` and `hour` here.

File: tv_apps/smart-clock/js/clock_component.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    this.setTransform('hour', 4);
+
+    // update again in one minute
+    this.timeouts.analog = setTimeout(this.updateAnalogClock.bind(this),
```
Comment (huchengtw-moz) [inline tv_apps/smart-clock/js/clock_component.js]: We have a shared clock.js which handles update per minute. How about use it?

File: tv_apps/smart-clock/js/clock_component.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+'use strict';
+(function(exports) {
+
+function ClockComponent() {}
+
+ClockComponent.prototype = {
```
Comment (huchengtw-moz) [inline tv_apps/smart-clock/js/clock_component.js]: This class is suitable to be a web component. It possible, please write it as web component.

File: tv_apps/smart-clock/manifest.webapp [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  "launch_path": "/widget.html",
+  "widgets": {
```
Comment (huchengtw-moz) [inline tv_apps/smart-clock/manifest.webapp]: The `widgetPages` array is required. Please find more information at https://wiki.mozilla.org/WebAPI/WidgetAPI . An example can be found at https://github.com/huchengtw-moz/gaia/tree/stingray-widget-api-v2 .

Comment (huchengtw-moz) [inline tv_apps/smart-clock/manifest.webapp]: The `previewImage` field is optional.

File: tv_apps/smart-clock/manifest.webapp [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+  "name": "Smart Clock",
```
Comment (huchengtw-moz) [inline tv_apps/smart-clock/manifest.webapp]: Please add icons field.

File: tv_apps/dashboard/manifest.webapp [hunk 8]
```diff
@@ -19,5 +19,10 @@
     "336": "/style/icons/dashboard_336.png",
     "672": "/style/icons/dashboard_672.png"
   },
-  "orientation": "landscape"
+  "orientation": "landscape",
+  "permissions": {
+    "browser": {},
+    "embed-widgets": {},
+    "webapps-manage": {}
```
Comment (huchengtw-moz) [inline tv_apps/dashboard/manifest.webapp]: Since we already require the permissions, we should include the widget.

Comment (huchengtw-moz) [inline tv_apps/dashboard/manifest.webapp]: The code to enumerate widgets can be found at:
https://github.com/huchengtw-moz/gaia/blob/58921c5dfebfc1998dd2ecd4a7308439782ac69e/dev_apps/homescreen-stingray/js/applications.js

You may trim them and bring widget handling to `shared/smart-screen/applications.js`.

Comment (rexboy7) [inline tv_apps/dashboard/manifest.webapp]: Got it.
Since I'm going to open a bug for embedding widget into dashboard, I'll include these codes there.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0493ec4fdf830276e223eaf3b4cea76e95e68620

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=92bc3afeae3df1eace34d54c603077e589ee5092

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0eaf0fb572e992e22f74afa40d19bdbdb708b8ef

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=98d6281535269ff1450e0538cd3b29e3597990f2

**Pull request merged**